tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -500,139 +498,203 @@ perform_vectorized_aggregation(DecompressChunkState *chunk_state)
-		Assert(decompress_all != NULL);
+			DecompressAllFunction decompress_all =
+				tsl_get_decompress_all_function(header->compression_algorithm);
+			Assert(decompress_all != NULL);
 
-		MemoryContext context_before_decompression =
-			MemoryContextSwitchTo(chunk_state->bulk_decompression_context);
+			MemoryContext context_before_decompression =
+				MemoryContextSwitchTo(chunk_state->bulk_decompression_context);
 
-		/* Fixme, type for column is wrong because its partial(sum(int4)) */
-		column_description->typid = 23;
+			/* Fixme, type for column is wrong because its partial(sum(int4)) */
+			column_description->typid = 23;
 
-		arrow = decompress_all(PointerGetDatum(header),
-							   column_description->typid,
-							   batch_state->per_batch_context);
+			arrow = decompress_all(PointerGetDatum(header),
+								   column_description->typid,
+								   batch_state->per_batch_context);
 
-		Assert(arrow != NULL);
+			Assert(arrow != NULL);
 
-		MemoryContextReset(chunk_state->bulk_decompression_context);
-		MemoryContextSwitchTo(context_before_decompression);
+			MemoryContextReset(chunk_state->bulk_decompression_context);
+			MemoryContextSwitchTo(context_before_decompression);
 
 #if defined(HAVE__BUILTIN_OP_OVERFLOW) && defined(USE_CUSTOM_SUM_IMPL)
 #define SUM_LANES 4
-		uint32 lane_result[] = { 0, 0, 0, 0 };
-		uint32 lane_buffer[] = { 0, 0, 0, 0 };
-		bool lane_overflow[] = { false, false, false, false };
+			uint32 lane_result[] = { 0, 0, 0, 0 };
+			uint32 lane_buffer[] = { 0, 0, 0, 0 };
+			bool lane_overflow[] = { false, false, false, false };
 
-		int cur_lane = 0;
+			int cur_lane = 0;
 
-		for (int i = 0; i < arrow->length; i++)
-		{
-			const bool arrow_isnull = !arrow_row_is_valid(arrow->buffers[0], i);
-
-			if (!arrow_isnull)
+			for (int i = 0; i < arrow->length; i++)
 			{
-				const uint32 arrow_value = ((uint32 *) arrow->buffers[1])[i];
-				lane_buffer[cur_lane] = arrow_value;
-				cur_lane = (cur_lane + 1) % 4;
-			}
+				const bool arrow_isnull = !arrow_row_is_valid(arrow->buffers[0], i);
 
-			if (cur_lane == 0)
-			{
-				for (int lane_selector = 0; lane_selector < SUM_LANES; lane_selector++)
+				if (!arrow_isnull)
 				{
-					lane_overflow[lane_selector] =
-						__builtin_add_overflow(lane_result[lane_selector],
-											   lane_buffer[lane_selector],
-											   &lane_result[lane_selector]);
+					const uint32 arrow_value = ((uint32 *) arrow->buffers[1])[i];
+					lane_buffer[cur_lane] = arrow_value;
+					cur_lane = (cur_lane + 1) % 4;
 				}
 
-				for (int lane_selector = 0; lane_selector < SUM_LANES; lane_selector++)
+				if (cur_lane == 0)
 				{
-					if (unlikely(lane_overflow[lane_selector]))
-						ereport(ERROR,
-								(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
-								 errmsg("bigint out of range")));
+					for (int lane_selector = 0; lane_selector < SUM_LANES; lane_selector++)
+					{
+						lane_overflow[lane_selector] =
+							__builtin_add_overflow(lane_result[lane_selector],
```

Comment:
Overflow checking is probably reducing the throughput. Some ways to address:
1) Just ignore it.
2) Use int64 accumulator for int32 input, it can't overflow inside one 1k row batch. Check the overflow at the end.
3) Don't check the overflow for each row. It's enough to OR it and check at some point later. Branching might prevent the vectorization.

Reply:
Since the partial sum of an int32 is a int64, i removed the overflow checks in the hot loop. 

I have three implementations for the sum function at the moment. All show more or less the same performance. Even when I force the use of SIMD data types, I see a slightly better performance.  For the time being, I will keep all three variations so that we can decide on which one to use during the review process.